motionplan/check.go

```diff
@@ -0,0 +1,430 @@
+		return err
+	}
+	// This should be done for any plan whose configurations are specified in relative terms rather than absolute ones.
+	// Currently this is only TP-space, so we check if the PTG length is >0.
+	// The solver frame will have had its PTGs filled in the newPlanManager() call, if applicable.
+	if sfPlanner.useTPspace {
+		return checkPlanRelative(checkFrame, executionState, worldState, fs, lookAheadDistanceMM, sfPlanner)
+	}
+	return checkPlanAbsolute(checkFrame, executionState, worldState, fs, lookAheadDistanceMM, sfPlanner)
+}
+
+func checkPlanRelative(
+	checkFrame referenceframe.Frame, // TODO(RSDK-7421): remove this
+	executionState ExecutionState,
+	worldState *referenceframe.WorldState,
+	fs referenceframe.FrameSystem,
+	lookAheadDistanceMM float64,
+	sfPlanner *planManager,
+) error {
+	plan := executionState.Plan()
+	startingInputs := plan.Trajectory()[0]
+	currentInputs := executionState.CurrentInputs()
+	currentPoseIF := executionState.CurrentPoses()[checkFrame.Name()]
+	wayPointIdx := executionState.Index()
+	sf := sfPlanner.frame
+
+	// Validate the given PoseInFrame of the relative frame. Relative frame poses cannot be given in their own frame, or the frame of
+	// any of their children.
+	// TODO(RSDK-7421): there will need to be checks once there is a real possibility of multiple, hierarchical relative frames, or
+	// something expressly forbidding it.
+	validateRelPiF := func(pif *referenceframe.PoseInFrame) error {
+		observingFrame := fs.Frame(pif.Parent())
+		// Ensure the frame of the pose-in-frame is in the frame system
+		if observingFrame == nil {
+			sfPlanner.logger.Errorf(
+				"pose of %s was given in frame of %s, but no frame with that name was found in the frame system",
+				checkFrame.Name(),
+				pif.Parent(),
+			)
+			return nil
+		}
+		// Ensure nothing between the PiF's frame and World is the relative frame
+		observingParentage, err := fs.TracebackFrame(observingFrame)
+		if err != nil {
+			return err
+		}
+		for _, parent := range observingParentage {
+			if parent.Name() == checkFrame.Name() {
+				return fmt.Errorf(
+					"pose of %s was given in frame of %s, bu current pose of checked frame must not be observed by self or child",
+					checkFrame.Name(),
+					pif.Parent(),
+				)
+			}
+		}
+		return nil
+	}
+
+	toWorld := func(pif *referenceframe.PoseInFrame, inputs map[string][]referenceframe.Input) (*referenceframe.PoseInFrame, error) {
+		// Check our current position is valid
+		err := validateRelPiF(pif)
+		if err != nil {
+			return nil, err
+		}
+		transformable, err := fs.Transform(inputs, pif, referenceframe.World)
+		if err != nil {
+			return nil, err
+		}
+		poseInWorld, ok := transformable.(*referenceframe.PoseInFrame)
+		if !ok {
+			// Should never happen
+			return nil, errors.New("could not convert transformable to a PoseInFrame")
+		}
+		return poseInWorld, nil
+	}
+
+	// Check out path pose is valid
+	stepEndPiF, ok := plan.Path()[wayPointIdx][checkFrame.Name()]
```

Comment:
there is a lot of new stuff going on here. I will come back to this in a following review.